flax/nnx/transforms/compilation.py

```diff
@@ -362,12 +408,34 @@ def __init__(
     functools.update_wrapper(self, fun)
     self.fun: tp.Callable[P, R] = fun
     kwarg_shardings = None
-    self.jax_in_shardings = jax.tree.map(
-      lambda x: extract.NodeStates.from_prefixes(x.shardings, metadata=x)
-      if isinstance(x, StateSharding)
-      else x,
-      in_shardings,
+    # Normalize static argnums/argnames to a tuple of positional indices.
+    self._static_positions: tuple[int, ...] = _normalize_static_argnums(
+      fun, static_argnums, static_argnames
     )
+
+    # Prepare jax_in_shardings for jax.jit by dropping entries that correspond
+    # to static positional arguments when a per-argument sequence is provided.
+    def _map_state_sharding(x):
+      return (
+        extract.NodeStates.from_prefixes(x.shardings, metadata=x)
+        if isinstance(x, StateSharding)
+        else x
+      )
+
+    def _drop_static_from_sequence(seq):
+      # Keep only shardings for dynamic (non-static) positions.
```

Comment:
Seems like this would be easier with a list comprehension one liner: 
`type(seq)([v for i,v in in enumerate(seq) if i not in self._static_positions])`.